drivers/adc/adc_demo/adc_demo.c

```diff
@@ -177,6 +178,50 @@ static bool get_next_ch_idx(uint32_t ch_mask, uint32_t last_idx,
 	return 1;
 }
 
+/***************************************************************************//**
+ * @brief Handles trigger: reads one data-set and writes it to the buffer.
+ *
+ * @param dev_data  - The iio device data structure.
+ *
+ * @return ret - Result of the handling procedure. In case of success, the size
+ * 				 of the written data is returned.
+*******************************************************************************/
+int32_t adc_demo_trigger_handler(struct iio_device_data *dev_data)
+{
+	struct adc_demo_desc *desc;
+	uint32_t k = 0;
+	uint32_t ch = -1;
+	uint16_t buff[TOTAL_ADC_CHANNELS];
+	static uint16_t i = 0;
+
+	if (!dev_data)
+		return -EINVAL;
+
+	desc = (struct adc_demo_desc *)dev_data->dev;
+
+	if(desc->ext_buff == NULL) {
+		int offset_per_ch = NO_OS_ARRAY_SIZE(sine_lut) / TOTAL_ADC_CHANNELS;
+		while(get_next_ch_idx(desc->active_ch, ch, &ch))
+			buff[k++] = sine_lut[(i + ch * offset_per_ch ) % NO_OS_ARRAY_SIZE(sine_lut)];
+		if (i == NO_OS_ARRAY_SIZE(sine_lut))
+			i = 0;
+		else
+			i++;
+
+		return iio_buffer_push_scan(dev_data->buffer, &buff[0]);
+	}
+
+	while(get_next_ch_idx(desc->active_ch, ch, &ch))
+		buff[k++] = ((uint16_t (*)[desc->ext_buff_len])(desc->ext_buff))[ch]
```

Comment:
Is the cast needed here? If not we should drop it to avoid obfuscating the code with complexity.

Reply:
yes, the cast is needed. I can also use the following approach if you think is less complex:
first get the pointer to the channel buffer:
`uint16_t* ch_buffer = (uint16_t*)(desc->ext_buff+(ch*desc->ext_buff_len*sizeof(uint16_t)/sizeof(ch_buffer)));`
then copy the data:
`buff[k++] = ch_buffer[i % desc->ext_buff_len];`